### DISCUSSION TITLE
Is there any plan for syslog-ng to use GDateTime to replace GTimeVal?

### DISCUSSION THREAD

**Discussion by ylse** [syslog-ng/syslog-ng#4292] [Q&A] [answered]
Hi, GTimeVal in glib-2.0 is being deprecated (still supported for some time but not recommended) since 2.62 because it is not year-2038-safe. It is recommended to use GDateTime to replace GTimeVal. I find there are some use of GTimeVal in syslog-ng. So I would like to know whether there is plan for the replacement or other solution for the future glib-2.0 without GTimeVal. Thank you in advance.

**Comment by bazsi** [comment, marked as answer]
We have our own UnixTime structure, which is used almost exclusively where a message timestamp is concerned:

```
typedef struct _UnixTime UnixTime;
struct _UnixTime
{
  gint64 ut_sec;
  guint32 ut_usec;

  /* zone offset in seconds, add this to UTC to get the time in local.  This
   * is just 32 bits, contrary to all other gmtoff variables, as we are
   * squeezed in space with this struct.  32 bit is more than enough for
   * +/-24*3600 */
  gint32 ut_gmtoff;
};
```

This is y2038 safe as we are using a 64 bit timestamp here, but this adds a timezone offset which might not be needed in all cases where we are currently using GTimeVal.

GDateTime is probably not the right fit, by a simple look at its implementation it seems to be pretty heavyweight, especially for our use-cases.

In places where the timezone information makes sense, I'd use our UnixTime structure, in other cases where it doesn't, I'd just use "struct timespec" or a 64bit time_t value.

Why do you ask? Could you help us to get rid of all those GTimeVals? It would be appreciated.

**Reply by ylse** [reply]
I created change to get rid of GTimeVal and functions using it here, https://github.com/syslog-ng/syslog-ng/pull/4479 , please check it when you have time.
